webserver/context-propagation/src/main/java/io/helidon/webserver/context/propagation/PropagationRecord.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (c) 2022 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.webserver.context.propagation;
+
+import io.helidon.config.Config;
+import io.helidon.config.metadata.Configured;
+import io.helidon.config.metadata.ConfiguredOption;
+import io.helidon.webserver.ServerRequest;
+
+/**
+ * Propagation record mapping header to classifier, may be a {@link java.lang.String} type, or an array of strings.
+ */
+public interface PropagationRecord {
+    /**
+     * Create a new fluent API builder to customize configuration.
+     *
+     * @return a new builder
+     */
+    static Builder builder() {
+        return new Builder();
+    }
+
+    /**
+     * Apply this record on the server request, reading the header and registering the context value.
+     *
+     * @param req server request
+     */
+    void apply(ServerRequest req);
+
+    /**
+     * Fluent API builder for {@link io.helidon.webserver.context.propagation.PropagationRecord}.
+     */
+    @Configured(description = "Context propagation record mapping HTTP header to context classifier.")
+    class Builder implements io.helidon.common.Builder<PropagationRecord> {
+        private String classifier;
+        private String headerName;
+        private String[] defaultValue;
+        private boolean isArray;
+
+        private Builder() {
+        }
+
+        @Override
+        public PropagationRecord build() {
+            if (defaultValue == null || defaultValue.length == 0) {
+                defaultValue = null;
+            }
+            if (classifier == null) {
+                classifier = headerName;
+            } else if (headerName == null) {
+                headerName = classifier;
+            }
+            if (headerName == null) {
+                throw new IllegalArgumentException("Either header name or classifier must be configured");
+            }
+
+            if (isArray) {
+                return new ArrayRecord(classifier, headerName, defaultValue);
```

Comment:
any considerations for record size limit / thresholds?

Reply:
we already have a limit on maximal size of headers, so this should not be needed. We may want to add it on client side (so we do not have too big headers).

Reply:
I will leave this for our future selves - if there is a need to limit the size (or length of an array), let's get an issue and fix it